homeassistant/components/sensor/phicomm.py

```diff
@@ -0,0 +1,232 @@
+
+    @property
+    def icon(self):
+        """Return the icon of the sensor."""
+        return self._icon
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit the value is expressed in."""
+        return self._unit
+
+    @property
+    def available(self):
+        """Return if the sensor data are available."""
+        data = self.data
+        return data and data.get('online') == '1'
+
+    @property
+    def state(self):
+        """Return the state of the device."""
+        return self.state_from_devs(self.phicomm.devs)
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return self.data if self._sensor_type == SENSOR_PM25 else None
+
+    @property
+    def should_poll(self):  # pylint: disable=no-self-use
+        """No polling needed."""
+        return False
+
+    @property
+    def data(self):
+        """Get data with current device index."""
+        devs = self.phicomm.devs
+        if devs and self._index < len(devs):
+            return devs[self._index].get('catDev')
+        return None
+
+    def state_from_devs(self, devs):
+        """Get state from Phicomm devs."""
+        if devs and self._index < len(devs):
+            return devs[self._index].get('catDev').get(self._sensor_type)
+        return None
+
+
+class PhicommData():
+    """Class for handling the data retrieval."""
+
+    def __init__(self, username, password, token_path, token, session, loop):
+        """Initialize the data object."""
+        self._username = username
+        self._password = password
+        self._token_path = token_path
+        self._session = session
+        self._loop = loop
+        self._token = token
+        self.devs = None
+        self.devices = None
+
+    async def async_update(self, time):
+        """Update online data and update ha state."""
+        old_devs = self.devs
+        await self.update_data()
+
+        tasks = []
+        for device in self.devices:
+            if device.state != device.state_from_devs(old_devs):
+                _LOGGER.debug('%s: => %s', device.name, device.state)
+                tasks.append(device.async_update_ha_state())
+
+        if tasks:
+            await asyncio.wait(tasks, loop=self._loop)
+
+    async def update_data(self):
+        """Update online data."""
+        try:
+            json = await self.fetch_data()
+            if json and json.get('error') != '0':
+                _LOGGER.debug("Reset token: error=%s", json.get('error'))
+                self._token = None
+                json = await self.fetch_data()
+
+            self.devs = json['data']['devs']
+            _LOGGER.debug("Get data: devs=%s", self.devs)
+        except (aiohttp.client_exceptions.ClientConnectorError, KeyError):
+            self.devs = None
+            import traceback
+            _LOGGER.error('Exception: %s', traceback.format_exc())
+
+    async def fetch_data(self):
```

Comment:
Data fetching and parsing should be moved to standalone library and published on pypi.